routers/api/packages/pypi/pypi.go

```diff
@@ -25,8 +25,8 @@ import (
 var normalizer = strings.NewReplacer(".", "-", "_", "-")
 var nameMatcher = regexp.MustCompile(`\A[a-zA-Z0-9\.\-_]+\z`)
 
-// https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions
-var versionMatcher = regexp.MustCompile(`^([1-9][0-9]*!)?(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*((a|b|rc)(0|[1-9][0-9]*))?(\.post(0|[1-9][0-9]*))?(\.dev(0|[1-9][0-9]*))?$`)
+// https://github.com/pypa/packaging/blob/0031046f7fad649580bc3127d1cef9157da0dd79/packaging/version.py#L113-L142
+var versionMatcher = regexp.MustCompile(`\Av?(?:(?:(?:[0-9]+)!)?(?:[0-9]+(?:\.[0-9]+)*)(?:[-_\.]?(?:a|b|c|rc|alpha|beta|pre|preview)[-_\.]?[0-9]*)?(?:(?:-[0-9]+)|(?:[-_\.]?(?:post|rev|r)[-_\.]?[0-9]*))?(?:[-_\.]?dev[-_\.]?[0-9]*)?)(?:\+(?:[a-z0-9]+(?:[-_\.][a-z0-9]+)*))?\z`)
```

Comment:
You could split it across lines and the regex could also be made slightly simpler by getting rid of some of the unneeded groups and changing the capture groups to non capturing groups (since we don't need to parse the version):

```suggestion
var versionMatcher = regexp.MustCompile(
    `v?([0-9]+!)?[0-9]+(\.[0-9]+)*` +                             // epoch + release segments
    `([-_\.]?(a|b|c|rc|alpha|beta|pre|preview)[-_\.]?[0-9]*)?` +  // pre-release
    `(-[0-9]+|[-_\.]?(post|rev|r)[-_\.]?[0-9]*)?` +               // post release
    `([-_\.]?dev[-_\.]?[0-9]*)?` +                                // dev release
    `(\+[a-z0-9]+([-_\.][a-z0-9]+)*)?`)                           // local version
```

Reply:
(And another pattern I have used in the past for managing complex regex is to have a link to an online regex test in a comment (something like regex101.com) that has a bunch of variations to match saved in a permalink in case someone comes back and wants to edit the regex.  Something like this: https://regex101.com/r/q3b2Xq/1)

Reply:
You're actually exchanging non-capturing groups for capturing groups (`(:?` are non-capturing groups). But it is much easier to read IMO... I don't know about the impact in performance though, because capturing groups allocate memory, while non-capturing does not, but it might be negligible.

Reply:
Ah yes got those swapped around.  We could add the :?s back to reduce memory.  This does have fewer groups overall as well

Reply:
Here is what that would look like:

```suggestion
var versionMatcher = regexp.MustCompile(
    `^v?(?:[0-9]+!)?[0-9]+(?:\.[0-9]+)*` +                             // epoch + release
    `(?:[-_\.]?(?:a|b|c|rc|alpha|beta|pre|preview)[-_\.]?[0-9]*)?` +   // pre-release
    `(?:-[0-9]+|[-_\.]?(?:post|rev|r)[-_\.]?[0-9]*)?` +                // post release
    `(?:[-_\.]?dev[-_\.]?[0-9]*)?` +                                 // dev release
    `(?:\+[a-z0-9]+(?:[-_\.][a-z0-9]+)*)?$`)                           // local version
```

Reply:
(and how that could look on regex101 https://regex101.com/r/cp5HSb/1)